app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,95 @@
+#
+# Handle signup or signin
+# from various oauth providers
+#
+# Heavily overlaps with Authentications controller
+#
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    create
+  end
+  def failure
+    flash[:error] = "Authentication failed."
+    redirect_to request.env['omniauth.origin'] || "/"
+  end
+
+  private
+
+  #
+  # Inspects the omniauth information
+  # and determines if we have an existing member 
+  # (to add authentication to)
+  # or if this is a new signup
+  #
+  # As a side affect, this method sets the @member_created
+  # variable
+  #
+  def member
+    @member_created = false
+
+    auth = request.env['omniauth.auth']
+
+    authentication = Authentication.where(provider: auth.provider, uid: auth.uid).first
+    if authentication && authentication.member.id.present?
+      member = authentication.member
+    end
+
+    member = Member.where(email: auth.info.email).first_or_create do |m|
+      m.email = auth.info.email
+      m.password = Devise.friendly_token[0,20]
+      m.tos_agreement = true
+      # TODO This has a reasonable chance of collision
+      m.login_name = auth.info.nickname || auth.info.email.split("@").first.gsub(/[^A-Za-z]+/, '_').underscore
+
+      m.skip_confirmation!
+
+      # TODO Assess this later, if we introduce separate modelling for user photos
+      # m.image = auth.info.image # assuming the user model has an image
+
+      @member_created = true
+    end
+    member.save!
+    
+    member
+  end
+
+  def create
+    auth = request.env['omniauth.auth']
+    @authentication = nil
+    if auth
+
+      name = ''
+      case auth['provider']
+      when 'twitter'
```

Comment:
I like that it looks like an easy future enhancement to allow sign in with twitter as well.